### ISSUE TITLE
Delete vectors for deleted points

### ISSUE THREAD

**Issue by timvisee** [qdrant/qdrant#2550]
**Describe the solution you'd like**
Points have vectors. When points are deleted, we should also mark their vectors as being deleted.

Though this isn't necessary for operation, doing this would allow our optimizers and query planner to handle situations with a lot of deleted vectors more efficiently.

**Additional context**
We have implemented this before in https://github.com/qdrant/qdrant/pull/1757

It turns out that this causes vectors to go missing on crash, so we disabled this behavior again in https://github.com/qdrant/qdrant/pull/2514

We should either fix the problems of vectors going missing when this is used, or find a different way of handling these deletions.

**Comment by fadilparves** [comment]
Hi,

Does that mean if we delete a Point it will not delete the vector that is associated with the Point?
I am interested in working on this feature, but it would be great if I could get some more information on it

Thanks :)

**Comment by timvisee** [comment]
Doing this properly turns out to be very hard as noted in the first comment. If Qdrant crashes, there are a few edge cases we cannot handle with our current architecture. It must be done properly to guarantee data consistency. We cannot fix this without significant changes in our architecture so we therefore mitigated this by removing propagating point deletes to vectors. I've opened this issue to look into it again in the future.

If the above doesn't clarify that yet; I do not recommend picking this up, especially as first issue. The concept seems simple, but the implementation is hard and has a lot of hidden foot guns (as we discovered months later).

Please note that the current state doesn't cause data inconsistency. If you delete a point, its vectors won't be considered for searches. Their data might remain on disk for some time though, but that isn't a burning issue.

**Comment by fadilparves** [comment]
Yep okay. Thanks for the reply @timvisee , I will look for other issue then or if you have any that you would recommend me to check then it would be great aswell. Thank you

**Comment by liebkne** [comment]
Any news of this?

**Comment by alex-cortenix** [comment]
> Doing this properly turns out to be very hard as noted in the first comment. If Qdrant crashes, there are a few edge cases we cannot handle with our current architecture. It must be done properly to guarantee data consistency. We cannot fix this without significant changes in our architecture so we therefore mitigated this by removing propagating point deletes to vectors. I've opened this issue to look into it again in the future.
> 
> If the above doesn't clarify that yet; I do not recommend picking this up, especially as first issue. The concept seems simple, but the implementation is hard and has a lot of hidden foot guns (as we discovered months later).
> 
> **Please note that the current state doesn't cause data inconsistency. If you delete a point, its vectors won't be considered for searches. Their data might remain on disk for some time though, but that isn't a burning issue.**

To clarify it a bit further: If a point was deleted, then its associated vector remains on a disk in 'orphaned' state. Will Qdrant delete it later automatically in background process, or it is a responsibility of a developer to take care about 'orphaned' vectors?

**Comment by timvisee** [comment]
> Will Qdrant delete it later automatically in background process, or it is a responsibility of a developer to take care about 'orphaned' vectors?

Qdrant will take care of this automatically, eventually. Collection data is stored in segments, once one of the optimizers pick a segment up, these _orphaned_ pionts are removed.

In fact, a developer does not have control over this. A developer could change optimizer thresholds, but I recommend against doing this for solely this purpose.

**Comment by ImTheDeveloper** [comment]
Stumbled upon this when looking at why my points count is 0 but vectors count is not.
![image](https://github.com/user-attachments/assets/4fe6b9bd-77df-44b7-aac9-fd6a62b65f10)
I assume therefore the 1877 vectors_count in my screenshot above are likely orphaned due to points_count 0 and eventually may get cleaned up, but arent part of search.
